maro/rl/learning/policy_manager.py

```diff
@@ -215,26 +217,58 @@ class DistributedPolicyManager(AbsPolicyManager):
     """
     def __init__(
         self,
-        policy_names: List[str],
+        create_policy_func_dict: Dict[str, Callable],
         group: str,
         num_hosts: int,
+        trainer_allocator: TrainerAllocator,
         log_dir: str = getcwd(),
         proxy_kwargs: dict = {}
     ):
         super().__init__()
-        self._policy_names = policy_names
+        self._policy_names = list(create_policy_func_dict.keys())
         peers = {"policy_host": num_hosts}
         self._proxy = Proxy(group, "policy_manager", peers, component_name="POLICY_MANAGER", **proxy_kwargs)
         self._logger = Logger("POLICY_MANAGER", dump_folder=log_dir)
+        self._trainer_allocator = trainer_allocator
+
+        self._policy2host, self._host2policies = self._trainer_allocator.allocate(
+            policy_name=self._policy_names, logger=self._logger)
+
+        self._logger.info("Spawning policy host processes")
+        self._state_cache = {}
+        self._policy_hosts = []
+        self._manager_end = {}
 
-        self._policy2host = {}
-        self._host2policies = defaultdict(list)
+        def _policy_host(name, create_policy_func, conn):
+            policy = create_policy_func(name)
+            _host_proxy = Proxy(group, "policy", peers, component_name=f"POLICY.{name}", **proxy_kwargs)
+            conn.send({"type": "init", "policy_state": policy.get_state()})
 
-        # assign policies to hosts
-        for i, name in enumerate(self._policy_names):
-            host_id = i % num_hosts
-            self._policy2host[name] = f"POLICY_HOST.{host_id}"
-            self._host2policies[f"POLICY_HOST.{host_id}"].append(name)
+            while True:
+                msg = conn.recv()
+                if msg["type"] == "learn":
+                    info_list = msg["rollout_info"]
+                    policy2host = msg["policy2host"]
+                    if isinstance(info_list[0], Trajectory):
+                        policy.distributed_learn(info_list, policy2host[name], name, _host_proxy)
```

Comment:
Is there an option to choose between learn and distributed_learn?

Reply:
Yes, sometimes `rollout_manager` may directly return the loss info, for example, using AC actors.